nameko/testing/utils.py

```diff
@@ -140,3 +140,70 @@ def find_free_port(host='127.0.0.1'):
     port = sock.getsockname()[1]
     sock.close()
     return port
+
+
+class ResourcePipeline(object):
+    """ Creates and destroys resources in background threads.
+
+    Creates up to `size` resources ahead of time so the caller avoids waiting
+    for lazy creation.
+    """
+
+    STOP = object()
+
+    def __init__(self, create, destroy, size=3):
+        if size == 0:
+            raise RuntimeError("Zero size would create unbounded resources")
+        self.ready = eventlet.Queue(maxsize=size)
+        self.trash = eventlet.Queue()
+
+        self.threads = []
+        self.create = create
+        self.destroy = destroy
+
+    def start(self):
+        self.running = True
+        self.threads.append(eventlet.spawn(self._create))
+        self.threads.append(eventlet.spawn(self._destroy))
+
+    def _create(self):
+        while self.running:
+            item = self.create()
+            self.ready.put(item)
+
+    def _destroy(self):
+        while True:
+            item = self.trash.get()
+            if item is ResourcePipeline.STOP:
+                break
+            self.destroy(item)
+
+    @contextmanager
+    def get(self):
+        item = self.ready.get()
+        try:
+            yield item
+        finally:
+            self.trash.put(item)
+
+    def shutdown(self):
+        self.running = False
+        while self.ready.qsize():
+            unused = self.ready.get()
+            # allow create thread to run if it is blocked waiting for capacity
+            # in the ready queue. otherwise we'll exit this loop before the
+            # final item is added to the ready queue and it won't be destroyed
+            eventlet.sleep()
```

Comment:
i'm still thinking about this but we might have a race here (e.g. if `create` is slow)